pinecone/management_client.go

```diff
@@ -0,0 +1,488 @@
+// on success. In case of failure, it returns an error describing the issue encountered.
+// It's important to check the returned error to understand the outcome of the request.
+//
+// Example:
+//
+//	projects, err := managementClient.ListProjects(ctx)
+//	if err != nil {
+//	    log.Fatalf("Failed to list projects: %v", err)
+//	}
+//	for _, project := range projects {
+//	    fmt.Printf("Project ID: %s, Name: %s\n", project.Id, project.Name)
+//	}
+func (c *ManagementClient) ListProjects(ctx context.Context) ([]*Project, error) {
+	res, err := c.restClient.ListProjectsWithResponse(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list projects: %w", err)
+	}
+
+	if res.JSON401 != nil {
+		return nil, fmt.Errorf("unauthorized: %s", res.JSON401.Error.Message)
+	}
+
+	if res.JSON500 != nil {
+		return nil, fmt.Errorf("internal server error: %s", res.JSON500.Error.Message)
+	}
+
+	if res.JSON4XX != nil {
+		return nil, fmt.Errorf("unknown error: %s", res.JSON4XX.Error.Message)
+	}
+
+	if res.JSON200 == nil || res.JSON200.Data == nil {
+		return nil, fmt.Errorf("unexpected response format or empty data")
+	}
+
+	projectList := make([]*Project, len(*res.JSON200.Data))
+	for i, p := range *res.JSON200.Data {
+		project := Project{
+			Id:   p.Id,
+			Name: p.Name,
+		}
+		projectList[i] = &project
+	}
+	return projectList, nil
+}
+
+// FetchProject retrieves a project by its ID from the management API. It makes a call to the
+// management plane's FetchProject endpoint and returns the project details.
+//
+// Parameters:
+// - ctx: A context.Context to control the request's lifetime.
+// - projectId: A string representing the unique identifier of the project to retrieve.
+//
+// Returns the project details on success or an error if the operation fails. Possible errors
+// include unauthorized access, project not found, internal server errors, or other HTTP client
+// errors.
+//
+// Example:
+//
+//	project, err := managementClient.FetchProject(ctx, "your_project_id_here")
+//	if err != nil {
+//	    log.Fatalf("Failed to fetch project: %v", err)
+//	}
+//	fmt.Printf("Project ID: %s, Name: %s\n", project.Id, project.Name)
+func (c *ManagementClient) FetchProject(ctx context.Context, projectId uuid.UUID) (*Project, error) {
+	resp, err := c.restClient.FetchProjectWithResponse(ctx, projectId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch project: %w", err)
+	}
+
+	switch resp.StatusCode() {
+	case http.StatusOK:
+		if resp.JSON200 != nil {
+			return &Project{
+				Id:   resp.JSON200.Id,
+				Name: resp.JSON200.Name,
+			}, nil
+		}
+	case http.StatusUnauthorized:
+		return nil, fmt.Errorf("unauthorized: %v", resp.JSON401)
+	case http.StatusNotFound:
+		return nil, fmt.Errorf("project not found: %v", resp.JSON404)
+	case http.StatusInternalServerError:
+		return nil, fmt.Errorf("internal server error: %v", resp.JSON500)
+	default:
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode())
+	}
```

Comment:
In the default case, is there any additional error message we can include from the response body? Otherwise the client will only see `"unexpected status code: 4xx"`, but any other details will be swallowed.

Reply:
good callout. Updated default on all calls to return the response body as well: 9692885